### PR TITLE
va-button: add submit functionality

### PR DESCRIPTION
**Pull request by rmessina1010**
## Chromatic

https://2379-submit-button--65a6e2ed2314f7b8f98609d8.chromatic.com

---
## Configuring this pull request
- [ ] Link to any related issues in the description so they can be cross-referenced.
- [ ] Add the appropriate version patch label (`major`, `minor`, `patch`, or `ignore-for-release`).
    - See [How to choose a version number](https://github.com/department-of-veterans-affairs/component-library#how-to-choose-a-version-number) for guidance.
    - Use `ignore-for-release` if files haven't been changed in a component library package. (ie. only Storybook files)
- [ ] DST Only: Increment the `/packages/core` version number if this will be the last PR merged before a release.
- [ ] Complete all sections below.
- [ ] Delete this section once complete

## Description
Closes [2379](https://github.com/department-of-veterans-affairs/vets-design-system-documentation/issues/2379)

## QA Checklist
- [ ] Component maintains 1:1 parity with design mocks
- [ ] Text is consistent with what's been provided in the mocks
- [ ] Component behaves as expected across breakpoints
- [ ] Accessibility expert has signed off on code changes (if applicable. If not applicable provide reason why)
- [ ] Designer has signed off on changes (if applicable. If not applicable provide reason why)
- [ ] Tab order and focus state work as expected
- [ ] Changes have been tested against screen readers (if applicable. If not applicable provide reason why)
- [ ] New components are covered by e2e tests; updates to existing components are covered by existing test suite
- [ ] Changes have been tested in vets-website using Verdaccio (if applicable. If not applicable provide reason why)

## Acceptance criteria
- [ ] QA checklist has been completed
- [ ] Screenshots have been attached that cover desktop and mobile screens

## Definition of done
- [ ] Documentation has been updated, if applicable
- [ ] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)

### CODE REVIEW

File: packages/web-components/src/components/va-button/va-button.tsx [hunk 1]
```diff
@@ -138,6 +156,12 @@ export class VaButton {
       return;
     }
     this.handleClick();
+    if (this.submit) {
+      // stops on click event
+      // e.preventDefault();
+      // e.stopPropagation();
```
Comment (jamigibbs) [inline packages/web-components/src/components/va-button/va-button.tsx:162]: Can you cleanup these commented out lines?

File: packages/web-components/src/components/va-button/va-button.tsx [hunk 2]
```diff
@@ -124,6 +129,19 @@ export class VaButton {
...
 
+  private handleSubmit() {
+    const theForm = this.el.closest('form');
+    const submitEvent = new CustomEvent('submit', {
+      bubbles: true,
+      cancelable: true,
+      composed: true,
+    });
+    theForm.dispatchEvent(submitEvent);
+    if (!this.dontSendForm) {
```
Comment (jamigibbs) [inline packages/web-components/src/components/va-button/va-button.tsx:140]: Can you explain here why this is necessary so that we can document the team or form's library needs?

File: packages/web-components/src/components/va-button/va-button.tsx [hunk 3]
```diff
@@ -80,6 +80,11 @@ export class VaButton {
    */
   @Prop({ reflect: true }) uswds?: boolean = true;
 
+  /**
+   * If `false`, buttons of type submit will run onsubmit call back, but not submit form.
+   */
+  @Prop() dontSendForm?: boolean = false;
```
Comment (jamigibbs) [inline packages/web-components/src/components/va-button/va-button.tsx:86]: Instead of creating a new prop, is it possible to leverage the existing `submit` prop to determine the different ways we need or want it to submit?

Comment (rmessina1010) [reply packages/web-components/src/components/va-button/va-button.tsx:86]: Changed the `submit` to a string,  'submit' or any truthy value will run the form's callback (`form.onsubmit`) and submit the data;  'prevent' will  run call back but not send data (essentially a faux `preventDefault`), 'skip' will  run call back but send data, but skip the form's callback.

**Comment by rmessina1010** [comment]
For general knowledge:
Due to complications with Stencil. The order of operations is:   form.onsubmit, button.onclick, and form.submit.

A workaround wold be to not use 'submit' and handle the process entirely  on the onclick event,  like this
```
    ... your onclick code here...
    const theForm = e.target.closest('form');
    const submitEvent = new CustomEvent('submit', {
       bubbles: true,
       cancelable: true,
       composed: true,
    });
    // the following calls the form.onsumit callback; omit if not wanted
    theForm.dispatchEvent(submitEvent);
    // the following submits the data/sends form; omit if not wanted
    theForm.submit(); 
```

File: packages/web-components/src/components/va-button/va-button.tsx [hunk 4]
```diff
@@ -66,9 +66,11 @@ export class VaButton {
   @Prop({ reflect: true }) secondary?: boolean = false;
 
   /**
-   * If `true`, the button will submit form data when clicked.
+   * A value of: `prevent` --will  triger the onsubmit callback on form, but not submit;
+   * `skip` --will submit but not trigger onsubmit callback; and `submit` or truthy -- will do both
+   * `false` acts a boolean false
```
Comment (Andrew565) [inline packages/web-components/src/components/va-button/va-button.tsx:71]: Should this be changed to "an empty string ('') acts as a boolean false"?

File: packages/web-components/src/components/va-button/va-button.tsx [hunk 5]
```diff
@@ -66,11 +66,11 @@ export class VaButton {
   @Prop({ reflect: true }) secondary?: boolean = false;
 
   /**
-   * A value of: `prevent` --will  triger the onsubmit callback on form, but not submit;
-   * `skip` --will submit but not trigger onsubmit callback; and `submit` or truthy -- will do both
-   * `false` acts a boolean false
+   *  Having this attribut present will set the type of this button as 'submit'.
```
Comment (Andrew565) [inline packages/web-components/src/components/va-button/va-button.tsx:69]: Missing the ending 'e' in attribute.

File: packages/web-components/src/components/va-button/va-button.tsx [hunk 6]
```diff
@@ -66,11 +66,11 @@ export class VaButton {
...
+   *  A value of: `prevent` --will  triger the onsubmit callback on form, but won't submit the form;
```
Comment (Andrew565) [inline packages/web-components/src/components/va-button/va-button.tsx:70]: Missing a 'g' in 'trigger'. Also grammatically should probably be "will trigger the onsubmit callback on *the* form".

**Review by Andrew565 — CHANGES_REQUESTED** [review]
Spelling and grammar errors.

File: packages/web-components/src/components/va-button/va-button.tsx [hunk 7]
```diff
@@ -66,9 +66,11 @@ export class VaButton {
   @Prop({ reflect: true }) secondary?: boolean = false;
 
   /**
-   * If `true`, the button will submit form data when clicked.
+   *  Having this attribut present will set the type of this button as 'submit'.
+   *  A value of: `prevent` --will  triger the onsubmit callback on form, but won't submit the form;
+   * `skip` --will submit the form but not trigger onsubmit callback;
    */
-  @Prop() submit?: boolean = false;
+  @Prop() submit?: string;
```
Comment (jamigibbs) [inline packages/web-components/src/components/va-button/va-button.tsx:73]: Will this break any existing implementations?

**Comment by Andrew565** [comment]
@rmessina1010 Just checking to make sure you got Jami and I's comments on this PR. :)

**Comment by FranECross** [comment]
@jamigibbs and @Andrew565 Just wondering if you know when this might release? We have several tickets blocked from implementation until it releases, and I'm updating statuses on them. Thanks! cc @randimays @chriskim2311

File: packages/web-components/src/components/va-button/va-button.tsx (hunk 7 above)
Comment (jamigibbs) [reply packages/web-components/src/components/va-button/va-button.tsx:73]: The reply left here is not publicly visible so here is the[ Slack convo](https://dsva.slack.com/archives/G01BJ3ESXL4/p1721234085222369?thread_ts=1721062755.570779&cid=G01BJ3ESXL4) and a screenshot of Ray's reply.

File: packages/web-components/src/components/va-button/va-button.tsx [hunk 8]
```diff
@@ -124,6 +126,32 @@ export class VaButton {
     return this.text;
   };
 
+  private handleSubmit() {
+    // const type = this.el.shadowRoot.querySelector('button')?.type;
+    // console.log(this.el.shadowRoot.querySelector('button'), 'XX');
+    // if (type !== 'submit') {
+    //   return;
```
Comment (jamigibbs) [inline packages/web-components/src/components/va-button/va-button.tsx:133]: Remove these comments.

File: packages/storybook/stories/va-button-uswds.stories.jsx [hunk 9]
```diff
@@ -31,15 +31,15 @@ const defaultArgs = {
 const Template = ({
   back,
   big,
-  'continue': _continue,
-  'disable-analytics': disableAnalytics,
+   _continue,
```
Comment (jamigibbs) [inline packages/storybook/stories/va-button-uswds.stories.jsx:34]: Why is this changed from `continue` to `_continue`?

File: packages/storybook/stories/va-button-uswds.stories.jsx [hunk 10]
```diff
@@ -106,3 +106,57 @@ Disabled.args = {
   disabled: true,
   text: "Disabled",
 };
+
+
+const Template2 = ({
```
Comment (jamigibbs) [inline packages/storybook/stories/va-button-uswds.stories.jsx:111]: Can you name this to something more descriptive than `Template2`?

File: packages/storybook/stories/va-button-uswds.stories.jsx [hunk 11]
```diff
@@ -106,3 +106,57 @@ Disabled.args = {
...
+  );
+}
+
+export const Submitted = Template2.bind(null);
+Submitted.args = {
+  ...defaultArgs,
+  onsub : (e)=>{ 
+    // e.preventDefault();
+    console.log(e.target, "I am submitted!!");
+    alert ("form onsubmit method fired!!--on form");
```
Comment (jamigibbs) [inline packages/storybook/stories/va-button-uswds.stories.jsx:153]: I think just one exclamation point is sufficient. Also when I submit, the Storybook page rerenders and shows this. Can this be stopped?

Screenshot 2024-07-23 at 8 43 16 AM

File: packages/storybook/stories/va-button-uswds.stories.jsx [hunk 12]
```diff
@@ -106,3 +106,57 @@ Disabled.args = {
...
+      />
+    </form>
+  );
+}
+
+export const Submitted = Template2.bind(null);
+Submitted.args = {
+  ...defaultArgs,
+  onsub : (e)=>{ 
+    // e.preventDefault();
```
Comment (jamigibbs) [inline packages/storybook/stories/va-button-uswds.stories.jsx:151]: Remove unnecessary comments please.

File: packages/web-components/src/components.d.ts [hunk 13]
```diff
@@ -265,9 +265,9 @@ export namespace Components {
          */
         "secondaryLabel"?: string;
         /**
-          * If `true`, the primary button will submit form data when clicked.
+          * Having this attribut present will set the type of this button as 'submit'. A value of: `prevent` --will  triger the onsubmit callback on form, but won't submit the form; `skip` --will submit the form but not trigger onsubmit callback;
```
Comment (jamigibbs) [inline packages/web-components/src/components.d.ts:268]: Typo: "attribut" should be "attribute".
Type: "--will" can just be "will"
Typo: "triger" should be "trigger"
Type: "callback on form" should be "callback on the form"
Typo: "trigger onsubmit callback;" shoudl be "trigger the onsubmit callback."

File: packages/web-components/src/components/va-button/va-button.tsx [hunk 14]
```diff
@@ -124,6 +126,32 @@ export class VaButton {
...
+    // }
+    if (this.submit === undefined) {
+      return;
+    }
+    const theForm = this.el.closest('form');
```
Comment (jamigibbs) [inline packages/web-components/src/components/va-button/va-button.tsx:138]: You should mention in the `submit` prop description that it requires a `form` element.

**Comment by rmessina1010** [comment]
> Spelling and grammar errors.

Corrected.

File: packages/web-components/src/components/va-button/va-button.tsx [hunk 15]
```diff
@@ -66,9 +66,13 @@ export class VaButton {
   @Prop({ reflect: true }) secondary?: boolean = false;
 
   /**
-   * If `true`, the button will submit form data when clicked.
+   * Having this attribute present will set the type of this button as 'submit'.
+   * The va-button element must be within a `form` element for this fuctionlity to take place
```
Comment (Andrew565) [inline packages/web-components/src/components/va-button/va-button.tsx:70]: functionality misspelled

**Review by Andrew565 — APPROVED** [review]
